krb5: call krb5_auth_cache_creds() if a pssword is available

sumit-bose opened the pull request:
The delayed online authentication feature is only available for password
authentication and krb5_auth_cache_creds() should only be called if a
password is available. Otherwise the error coded returned by
krb5_auth_cache_creds() will prevent proper offline authentication with
other methods e.g. Smartcard authentication.

Resolves https://pagure.io/SSSD/sssd/issue/3564

fidencio commented:
@amitkumar50, thanks for the code-review.

I'd like to ask you (or any other code reviewer) to set up a similar environment than the use where the bug was found, reproduce the issue without this patch, apply  the patch and verify that the issue is gone and only then give an ACK (would be nice to run our downstream tests as a well, but that's a different topic) ... otherwise we may end up introducing issues/regressions in our code.

lslebodn commented on src/providers/krb5/krb5_auth.c:
```diff
@@ -1138,7 +1138,9 @@ static void krb5_auth_done(struct tevent_req *subreq)
-                            KRB5_STORE_PASSWORD_IF_OFFLINE)) {
+                            KRB5_STORE_PASSWORD_IF_OFFLINE)
+                && sss_authtok_get_type(pd->authtok)
+                            == SSS_AUTHTOK_TYPE_PASSWORD) {
```
There was added a check for `authtok` type password. Similar check is also in `krb5_auth_cache_creds` at the beginning.

```
247 static void krb5_auth_cache_creds(struct krb5_ctx *krb5_ctx,
 248                                   struct sss_domain_info *domain,
 249                                   struct confdb_ctx *cdb,
 250                                   struct pam_data *pd, uid_t uid,
 251                                   int *pam_status, int *dp_err)
 252 {
 253     const char *password = NULL;
 254     errno_t ret;
 255 
 256     if (sss_authtok_get_type(pd->authtok) != SSS_AUTHTOK_TYPE_PASSWORD) {
 257         DEBUG(SSSDBG_MINOR_FAILURE,
 258               "Delayed authentication is only available for password "
 259               "authentication (single factor).\n");
 260         return;
 261     }
```

After this change that part of code will be dead in `krb5_auth_cache_creds`. Should we remove it?
Or should we set `pam_status` and `dp_err` the same as in else branch,
Because debug message in `krb5_auth_cache_creds` is a little bit more suitable in context then message in else branch *Backend is marked offline, retry later!*

lslebodn commented:
@sumit-bose,

I would like to apologize for long pending review. I added inline comment.

sumit-bose commented:
Hi @lslebodn,

Thank you for the review. I removed the check from krb5_auth_cache_creds() and moved the debug message to the error message of sss_authtok_get_password().

The 'Backend is marked offline, retry later!' messages is ok here because we just want to return the offline state here. Always saying that delayed authentication which is not enabled by default is not available might be irritating.

lslebodn commented:
Thank you for update.
I'll run some krb5 tests for sanity checking.

So far LGTM

lslebodn commented:
Sanity test with krb5/ad/ipa passed

ACK

lslebodn commented:
master:
* a87658e5382a6ad119058d22b118a29eaae7a365

sssd-1-14:
* 3546e6c0be2f85c9e1828237000a376a188e43ac

sssd-1-13:
* 6b900667cd1541809dbaaf9603f7e7785e6f78f9
